corehq/apps/reports/static/reports/v2/js/datagrid/columns.js

```diff
@@ -5,82 +5,156 @@
+    'underscore',
+    'reports/v2/js/datagrid/filters',
 ], function (
     $,
-    ko
+    ko,
+    _,
+    filters
 ) {
     'use strict';
 
-    var columnModel = function (data) {
+    var columnModel = function (data, availableFilters) {
         var self = {};
 
         self.title = ko.observable(data.title);
-        self.slug = ko.observable(data.slug);
+        self.name = ko.observable(data.name);
         self.width = ko.observable(data.width || 200);
 
+        self.appliedFilters = ko.observableArray(_.map(data.appliedFilters, function (filterData) {
+            var filterModel = filters.appliedColumnFilterModel(filterData);
+            if (availableFilters) {
+                filterModel.filter(ko.utils.arrayFirst(availableFilters(), function (item) {
+                    return item.name() === filterModel.filter().name() && item.filterType() === filterModel.filter().filterType();
+                }));
+            }
+            return filterModel;
+        }));
+
         self.unwrap = function () {
             return ko.mapping.toJS(self);
         };
 
+        self.getFilters = function () {
+            return _.map(self.appliedFilters(), function (appliedFilter) {
+                return {
+                    propertyName: self.name(),
+                    filterType: appliedFilter.filter().filterType(),
+                    filterName: appliedFilter.filter().name(),
+                    filterValue: appliedFilter.value() || "",
+                };
+            });
+        };
+
         return self;
     };
 
     var editColumnController = function (options) {
         var self = {};
 
-        self.slugEndpoint = options.slugEndpoint;
-        self.reportContext = options.reportContext;
+        self.endpoint = options.endpoint;
+
+        self.availableFilters = ko.observableArray(_.map(options.availableFilters, function (data) {
+            return filters.columnFilterModel(data);
+        }));
 
-        self.slugOptions = ko.observableArray();
+        self.defaultFilterType = options.availableFilters[0].filterType;
+        self.appliedFilterType = ko.observable();
+        self.filterTitleByType = _.object(_.map(options.availableFilters, function (data) {
+            return [data.filterType, data.filterTypeTitle];
+        }));
+        self.filterTypeOptions = ko.observableArray(_.keys(self.filterTitleByType));
+
+        self.filterOptions = ko.computed(function () {
+            return _.filter(self.availableFilters(), function (filter) {
+                return filter.filterType() === self.appliedFilterType();
+            });
+        });
+
+        self.columnNameOptions = ko.observableArray();
 
         self.oldColumn = ko.observable();
         self.column = ko.observable();
         self.isNew = ko.observable();
+        self.hasFilterUpdate = ko.observable(false);
+
+        self.init = function (reportContextObservable) {
+            self.reportContext = reportContextObservable;
+        };
```

Comment:
Would it be possible to pass `reportContextObservable` to the constructor rather than having a separate `init` function?  Do you need the uninitialized version somewhere?

Reply:
i wanted to have a clear way to specify that `reportContext` is an observable and have it passed consistently to the models that utilize it. I also wanted to generate it after the `columns` are loaded in `datagrid`'s `init` method